be/src/exec/pipeline/aggregate/aggregate_streaming_sink_operator.cpp

```diff
@@ -89,63 +89,190 @@ Status AggregateStreamingSinkOperator::_push_chunk_by_force_preaggregation(Chunk
+ *
+ * SELECTIVE_PREAGG state aggregates continuous_limit chunks, then shifting to ADJUST state.
+ */
+Status AggregateStreamingSinkOperator::_push_chunk_by_auto(ChunkPtr chunk, const size_t chunk_size) {
+    size_t allocated_bytes = _aggregator->hash_map_variant().allocated_memory_usage(_aggregator->mem_pool());
+    const size_t continuous_limit = _auto_context.get_continuous_limit();
+    switch (_auto_state) {
+    case AggrAutoState::INIT_PREAGG: {
+        size_t real_capacity =
+                _aggregator->hash_map_variant().capacity() - _aggregator->hash_map_variant().capacity() / 8;
+        size_t remain_size = real_capacity - _aggregator->hash_map_variant().size();
+        bool ht_needs_expansion = remain_size < chunk_size;
+        _auto_context.init_preagg_count++;
+        if (!ht_needs_expansion ||
+            _aggregator->should_expand_preagg_hash_tables(_aggregator->num_input_rows(), chunk_size, allocated_bytes,
+                                                          _aggregator->hash_map_variant().size())) {
+            // hash table is not full or allow to expand the hash table according reduction rate
             SCOPED_TIMER(_aggregator->agg_compute_timer());
-            RETURN_IF_ERROR(_aggregator->compute_batch_agg_states(chunk.get(), chunk_size));
+            TRY_CATCH_BAD_ALLOC(_aggregator->build_hash_map(chunk_size));
+            if (_aggregator->is_none_group_by_exprs()) {
+                RETURN_IF_ERROR(_aggregator->compute_single_agg_state(chunk.get(), chunk_size));
+            } else {
+                RETURN_IF_ERROR(_aggregator->compute_batch_agg_states(chunk.get(), chunk_size));
+            }
+
+            _mem_tracker->set(_aggregator->hash_map_variant().reserved_memory_usage(_aggregator->mem_pool()));
+            TRY_CATCH_BAD_ALLOC(_aggregator->try_convert_to_two_level_map());
+
+            COUNTER_SET(_aggregator->hash_table_size(), (int64_t)_aggregator->hash_map_variant().size());
+            break;
         } else {
-            // middle cases, first aggregate locally and output by stream
-            {
-                SCOPED_TIMER(_aggregator->agg_compute_timer());
-                RETURN_IF_ERROR(_aggregator->compute_batch_agg_states_with_selection(chunk.get(), chunk_size));
+            _auto_state = AggrAutoState::ADJUST;
+            _auto_context.adjust_count = 0;
+            LOG(INFO) << "auto agg: " << _auto_context.get_auto_state_string(AggrAutoState::INIT_PREAGG) << " "
+                      << _auto_context.init_preagg_count << " -> " << _auto_context.get_auto_state_string(_auto_state);
+        }
+    }
+    case AggrAutoState::ADJUST: {
+        _auto_context.adjust_count++;
+        {
+            SCOPED_TIMER(_aggregator->agg_compute_timer());
+            TRY_CATCH_BAD_ALLOC(_aggregator->build_hash_map_with_selection(chunk_size));
+        }
+
+        size_t agg_count = SIMD::count_zero(_aggregator->streaming_selection());
```

Comment:
I doubt this `agg_count` may not exact... eg, for a new chunk:

keys in hash_map already exists: 1, 2, 2, 3
new keys from chunk: 5, 5, 5, 5, 

all `5` is not found... 

agg_count is still 3

another cases: new keys from chunk: 5, 6, 7, 8
agg_count is still 3

use `agg_count` to determine whether low/high reduction may disturb?

Reply:
agg_count means how many found keys of a chunk from a hash table.
the first case, agg_count = 0, so the chunk(5,5,5,5) is passed.
the second case is also passed.